### PR TITLE
Use ProviderConfig instead of Provider

### PR DESCRIPTION
**Pull request by muvaf**
### Description of your changes

* Updates all docs & examples to use `ProviderConfig` type and `providerConfigRef` field.
* Adds new `ProviderConfig` type at v1beta1 level for Kubernetes native providers like provider-rook.

Part of https://github.com/crossplane/crossplane/issues/1684

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: cluster/examples/cache/rediscluster/aws/resource-class.yaml [hunk 1]
```diff
@@ -7,7 +7,7 @@ metadata:
     example: "true"
 specTemplate:
   writeConnectionSecretsToNamespace: crossplane-system
-  providerRef:
+  providerConfigRef:
```
Comment (hasheddan) [inline cluster/examples/cache/rediscluster/aws/resource-class.yaml:10]: Did we update classes to use `providerConfigRef`? I thought we only did managed resources

Comment (muvaf) [reply cluster/examples/cache/rediscluster/aws/resource-class.yaml:10]: Oh, that's right. Will update the PR. Thanks for catching!

Comment (muvaf) [reply cluster/examples/cache/rediscluster/aws/resource-class.yaml]: @hasheddan updated

File: design/one-pager-strongly-typed-class.md [hunk 2]
```diff
@@ -186,7 +186,7 @@ specTemplate:
     - sg-ab1cdefg
     - sg-05adsfkaj1ksdjak
   size: 20
-  providerRef:
+  providerConfigRef:
```
Comment (hasheddan) [inline design/one-pager-strongly-typed-class.md:189]: Looks like there are still a number of classes where these changes don't need to be made

Comment (muvaf) [reply design/one-pager-strongly-typed-class.md]: Now it should be OK.

File: docs/contributing/services_developer_guide.md [hunk 3]
```diff
@@ -607,7 +607,7 @@ feedback you may have about the development process!
 [managed resource]: concepts.md#managed-resource
 [dynamic provisioning]: concepts.md#dynamic-and-static-provisioning
 [`CloudMemorystoreInstance`]: https://github.com/crossplane/provider-gcp/blob/85a6ed3c669a021f1d61be51b2cbe2714b0bc70b/apis/cache/v1beta1/cloudmemorystore_instance_types.go#L184
-[`Provider`]: https://github.com/crossplane/provider-gcp/blob/85a6ed3c669a021f1d61be51b2cbe2714b0bc70b/apis/v1alpha3/types.go#L41
+[`ProviderConfig`]: https://github.com/crossplane/provider-gcp/blob/85a6ed3c669a021f1d61be51b2cbe2714b0bc70b/apis/v1alpha3/types.go#L41
```
Comment (hasheddan) [inline docs/contributing/services_developer_guide.md:610]: I think this link needs updating

**Pull request merged**